streaming/src/queue/queue_manager.h

```diff
@@ -0,0 +1,153 @@
+                  bool raw = false /*only for test*/);
+
+ private:
+  std::shared_ptr<QueueManager> manager_;
+};
+
+/// Interface for Streaming Reader
+class QueueReader {
+ public:
+  QueueReader(std::shared_ptr<QueueManager> manager) : manager_(manager) {}
+
+  virtual ~QueueReader();
+
+  bool CreateQueue(const ObjectID &queue_id, const ActorID &actor_id,
+                   const ActorID &peer_actor_id, uint64_t start_seq_id);
+  void GetQueueItem(const ObjectID &queue_id, uint8_t *&data, uint32_t &data_size,
+                    uint64_t &seq_id, uint64_t timeout_ms = -1);
+  void NotifyConsumedItem(const ObjectID &queue_id, uint64_t seq_id);
+
+ private:
+  std::shared_ptr<QueueManager> manager_;
+};
+
+/// QueueManager manages upstream and downstream queues for worker. Threre should be
+/// only one QueueManager instance for a worker. Singleton.
+/// QueueManager holds a boost.asio io_service (queue thread). For upstream, queue
+/// thread handle events from main thread, for example, notify/pull. For downstream,
+/// queue thread receive and handle all events from main thread, to make main thread
+/// return quickly.
+/// Interface for internal threads.
+class QueueManager {
+ public:
+  QueueManager(const ActorID &actor_id)
+      : actor_id_(actor_id), queue_dummy_work_(queue_service_) {
+    Init();
+  }
+
+  virtual ~QueueManager() {
+    Stop();
+  }
+
+  /// TODO: Remove singleton if we have a proper initialization sequence.
+  static std::shared_ptr<QueueManager> GetInstance(const ActorID &actor_id) {
+    if (nullptr == queue_manager_) {
+      queue_manager_ = std::make_shared<QueueManager>(actor_id);
+    }
+
+    return queue_manager_;
+  }
+
+  std::shared_ptr<WriterQueue> CreateUpstreamQueue(const ObjectID &queue_id,
+                                             const ActorID &actor_id,
+                                             const ActorID &peer_actor_id, uint64_t size);
+  std::shared_ptr<ReaderQueue> CreateDownstreamQueue(const ObjectID &queue_id,
+                                               const ActorID &actor_id,
+                                               const ActorID &peer_actor_id);
+  bool UpstreamQueueExists(const ObjectID &queue_id);
+  bool DownstreamQueueExists(const ObjectID &queue_id);
+
+  /// Wait all queues in queue_ids vector ready, until timeout.
+  void WaitQueues(const std::vector<ObjectID> &queue_ids, int64_t timeout_ms,
+                  std::vector<ObjectID> &failed_queues, QueueType type);
+
+  void UpdateUpActor(const ObjectID &queue_id, const ActorID &actor_id);
+  void UpdateDownActor(const ObjectID &queue_id, const ActorID &actor_id);
+
+  void AddOutTransport(const ObjectID &actor_id, std::shared_ptr<Transport> transport);
+  std::shared_ptr<Transport> GetOutTransport(const ObjectID &actor_id);
+
+  std::shared_ptr<LocalMemoryBuffer> OnCheckQueue(
+      std::shared_ptr<CheckMessage> check_msg);
+  void OnCheckQueueRsp(std::shared_ptr<CheckRspMessage> check_rsp_msg);
```

Comment:
> what does rsp stand for?

for upstream and downstream queues, it may check whether peer is ready or not. the checking process is synchronous, the peer actor should send a response to caller. `rsp` stands for `response`

Reply:
> what does rsp stand for?

OnCheckQueueRsp has been removed when we refactor code.